### PR TITLE
universe: Add Terraform package

### PR DESCRIPTION
**Pull request by efoncubierta**
Signed-off-by: Ezequiel Foncubierta <ezequiel@foncubierta.com>

I know there is another contribution for a Terraform package ([#2110](https://github.com/dagger/dagger/pull/2110)), but this one we are currently using and implementing at work, and supports a lot more features. The approach is similar to Nico's implementation, so feel free to merge it into his package, or the other way around, if necessary.

### Implementation details

There is a `terraform.#Run` action to wrap the Terraform command and run it within the official Terraform container image, from Hashicorp. You can set the Terraform command you want to run (i.e. plan, apply), the arguments (i.e. -var-file, -var), and environment variables. For common environment variables or arguments, there is a field in `terraform.#Run` you can use. For example, the `workspace` fields will become a `TF_WORKSPACE` environment variable.

For each particular Terraform command (i.e. plan, apply) there is an action (i.e. `terraform.#Plan`) that extends `terraform.#Run`. These actions override the command field and define additional fields when necessary. For example, a `planFile` field.

For things that are not "yet supported", you can always fall back to the `terraform.#Run` action.

The `terraform.#Run` actions have two outputs:

-   `output`: A `dagger.#FS` with the modified Terraform workspace.
-   `outputImage`: The `docker.#Image` that was used to run the command.

### Not yet supported

-   Authentication in Terraform Cloud: [https://www.terraform.io/cli/auth](https://www.terraform.io/cli/auth)
-   Explicit actions (you can still use `terraform.#Run` though):
    - to import infrastructure: [https://www.terraform.io/cli/import](https://www.terraform.io/cli/import)
    - to inspect infrastructure: [https://www.terraform.io/cli/inspect](https://www.terraform.io/cli/inspect)
    - to manipulate state: [https://www.terraform.io/cli/state](https://www.terraform.io/cli/state)
    - to manage plugins: [https://www.terraform.io/cli/plugins](https://www.terraform.io/cli/plugins)
-   Use a container as input for `terraform.#Run` so you can chain multiple commands within the same container?

### CODE REVIEW

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/workspace.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/workspace.cue:3]: That should be a public definition

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/workspace.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/workspace.cue:7]: You could interpolate instead of `+`?

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:9]: That should be public

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:14]: You should use a stable version instead of `latest`, 
And create a `terraform.#Image` definition would be a good idea to predefine an image or help user provide his own instead of giving a container to pull.
You can take inspiration from [`go.#Image`](https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/go/image.cue)

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:23]: It's already empty by default

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:26]: It's already empty by default

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:56]: Nit: You could use `mounts` in `docker.#Run` instead of copy

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:72]: Nit: You could use `exports` from `docker.#Run` instead of an intermediate steps to retrieve directory

**Review by helderco — APPROVED** [review]
Thanks!

You should comment all your public fields so it gets picked up by the CLI:
https://docs.dagger.io/1226/coding-style

Also, please squash your commits:
https://docs.dagger.io/1227/contributing

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:51]: Could be `terraform.#Image` or one provides by the user :D

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:97]: Why you just do not set those value instead of an `if` statement?
It should not break if `TF_WORKSPACE` is set to `default`

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:80]: Nit: explicit comment

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/plan.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/plan.cue:4]: Should be directly the type of `planFile`, you do not need an intermediate definition for that

**Review by TomChv — COMMENTED** [review]
Thank for your contribution, it's an amazing work!
I really like your implementation, it's a good one.

I left some comments to make it even more awesome :rocket:
I will take pleasure to follow up your work on that PR :D

**Comment by TomChv** [comment]
> The approach is similar to Nico's implementation, so feel free to merge it into his package, or the other way around, if necessary.

Since packages are experimental, it's okay to have duplication.
We will make a choice when we will port a terraform package in the "official" universe

**Comment by efoncubierta** [comment]
I amended most of the requested changes, but using mount instead of `docker.#Copy`. The `/src` directory is empty when I use mount.

I am getting the following error when running `dagger do lint`. I don't know exactly what's going on here.

```
1:49PM FTL failed to execute plan: task failed: actions.lint.cue._dag."4"._exec: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1
```

Re squashing commit. I was able to do it only on commits after the rebase from `main`.

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [inline pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:80]: You could directly use `docker.#Run` outside `docker.#Build`, I'll try your package and show you ;)

**Review by TomChv — APPROVED** [review]
It looks really good! 

I'll play with your package to understand why you can't simplify the exports.
However, it's already mergeable, we can improve package's internal later ;)

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (efoncubierta) [reply pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:80]: For completeness, this is what I tried and didn't work

```
#Run: {
  ...

  _image: #Image

  // Run command within a container
  _run: docker.#Run & {
    input: *_image.output | docker.#Image
    workdir: "/src"
    command: {
      name: cmd
      args: _thisCmdArgs
    }
    mounts: src: {
      dest: "/src"
      contents: source
    }
    env: _thisEnv
  }

  ...
}
```

**Comment by TomChv** [comment]
> Really looking forward to merge this. What's missing? Anything we can help with?

Fixing lint and another approval from me after a manual test of the package, I'll do it tomorrow

**Comment by aluzzardi** [comment]
Thanks @TomChv!

File: pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue
Comment (TomChv) [reply pkg/universe.dagger.io/x/ezequiel@foncubierta.com/terraform/run.cue:80]: Oh you are right, you are updating `/src` and outputting it for further operation so yes, you did great!

**Comment by TomChv** [comment]
I manually tested the package! It's really good :rocket:

We can merge that package just after lint error are fixed.

@efoncubierta You can just run `make cuefmt` and then amend your last commit to apply lint.
We will merge your package just after that

**Comment by TomChv** [comment]
@efoncubierta I fixed your linter error, just squash all your commits (do not forget to pull, I've fix your lint error) and we will merge that one!

**Comment by efoncubierta** [comment]
@TomChv I squashed the commits. I can't squash the two commits before the merge from the main branch.

**Comment by TomChv** [comment]
@aluzzardi Ready for merge!

**Pull request merged**
